docs/proposals/002-refresh-certificates.md

```diff
@@ -0,0 +1,371 @@
+will need to be approved and signed. This can be used by the CLI to print a
+helpful message and request that the certificates be signed.
+
+This method call is purely informative and does not have any side-effect on the
+cluster.
+
+### `POST /refresh-certs/run`
+
+```go
+type RefreshCertificatesRunRequest struct {
+  // Seed must match the seed from the received RefreshCertificatesInitResponse.
+  Seed int
+  // ExpirationDuration is the duration of the requested certificates.
+  ExpirationDuration time.Duration
+}
+
+type RefreshCertificatesRunResponse struct {
+  // ExpirationDuration is the duration of the new certificates.
+  ExpirationDuration time.Duration
+}
+```
+
+#### Worker nodes
+
+Worker nodes will create the required CSRs, and wait for them to be signed. This
+comprises of two steps: first they must be approved (see below), and then the
+csrsigning controller will sign them. The worker node will block until the CSRs
+are dealt with:
+
+- If any CSR becomes Failed or Denied, then we fail and return the result.
+- If any CSR does not become Approved or Issued in time, we time out.
+
+After getting the new certificates, we apply them on the node, do any service
+restarts that may be required and return response.
+
+#### Control plane nodes
+
+Control plane nodes can ignore the seed argument. They already have access to
+the CA certificates and keys, therefore can simply generate and sign new
+certificates, apply them on the node, do any service restarts necessary and
+return the response.
+
+## CLI Changes
+<!--
+This section MUST mention any changes to the k8s CLI, e.g. new arguments,
+different outputs.
+-->
+
+### `k8s refresh-certs` command
+
+#### For worker nodes
+
+```
+root@worker:~$ k8s refresh-certs --ttl 30d --timeout 30s
+The following CertificateSigningRequests should be approved. Run the following
+commands on any of the control plane nodes of the cluster:
+
+  $ k8s kubectl certificate approve k8sd-seed-worker-kubelet-client
+  $ k8s kubectl certificate approve k8sd-seed-worker-kubelet-serving
+  $ k8s kubectl certificate approve k8sd-seed-worker-kube-proxy-client
+
+Waiting for certificates to be created....
+
+<-- CSRs are approved -->
+
+Certificates have been refreshed, and will expire at $date.
+```
+
+#### For control plane nodes
+
+```
+root@node:~$ k8s refresh-certs --ttl 30d --timeout 30s
+Certificates have been refreshed, and will expire at $date.
+```
+
+## Database Changes
+<!--
+This section MUST mention any changes required in the k8sd database schema or
+internal types.
+-->
+none
+
+## Configuration Changes
+<!--
+This section MUST mention any new configuration options or service arguments
+that are introduced.
+-->
+
+The changes include a `csrsigning` controller, which watches the cluster for
+`CertificateSigningRequest` objects coming from the worker nodes. After they are
+approved, it will sign the certificates.
+
+### k8sd `--disable-csrsigning-controller`
```

Comment:
if we go with the `CSRSigning` cluster config, we could also provide this as a bootstrap/config option. If a user really don't want this feature, it would feeld cumbersome to change this on every node. We would still have this flag if the user wants to disable the controller just on one node.